storage/norcow.c

```diff
@@ -100,6 +117,85 @@ static secbool norcow_write(uint8_t sector, uint32_t offset, uint32_t prefix,
   ensure(flash_lock_write(), NULL);
   return sectrue;
 }
+#else
+static secbool norcow_write(uint8_t sector, uint32_t offset, uint16_t key,
+                            const uint8_t *data, uint16_t len) {
```

Comment:
The else branch is somewhat problematic because of the udpate_bytes mechanism - can't write the key until enough data has been pushed through update_bytes, but update_bytes will not find the item in storage until the key is written. 

We could do something like this in the else branch:

- First quadword:
     - 2 bytes len
     - 2 bytes key
- Second quadword:
     - 1 byte deleted flag 
     - 15 bytes data

Is it ok to have the key remain in the first QW after deletion?

Reply:
There are additional issues with the else branch (even with the updated version):
- once you start streaming the bytes, you have to finish before streaming some other key, otherwise it fails (this is not broken anywhere at this moment, but its a new limitation that one needs to be aware of)
- the current implementation of storage's `secequal32` expects the data aligned to 32bits. This is also breaking boundaries between storage and norcow as its making an assumption about how the data are organized. We can either remove the asserts (our current MCU's can handle unaligned access with some performance penalty, which i deem acceptable at this situation), or we can perhaps use `secequal` instead. 

With that said, here is implementation of this optimization and also the update_bytes buffering: https://github.com/trezor/trezor-firmware/commit/51666699faa5e6735147fa7927cddd37a8d81877

Note that test_norcow_qw are currently failing as i didn't to waste more time until we decide that this is a way to go. Other tests are ok. Also the data structure is incompatible with older storages as mentioned below: we could either have two variants or do a storage upgrade.

Reply:
In subsequent commits https://github.com/trezor/trezor-firmware/pull/3228/commits/d278e20f8f1181cea9a8f289b1986f2f60c5968c and https://github.com/trezor/trezor-firmware/pull/3228/commits/d3de9942eb080b1ab29f25f22906e53185e3c16c
I have merged the pin logs behavior in storage.c: now it doesn't use `update_word` but always calls norcow_set, as this function is able to determine what is updatable and what isn't.
And counter implementation is moved to norcow.c, because how it is implemented is dependent on underlying hw. 

With these changes, storage.c is now independent on how norcow.c is implemented